apps/mobile-wallet/src/features/auto-lock/useAutoLock.tsx

```diff
@@ -0,0 +1,114 @@
+
+import { useAppDispatch, useAppSelector } from '~/hooks/redux'
+
+interface UseAutoLockProps {
+  unlockApp: () => Promise<void>
+  onAuthRequired: () => void
+}
+
+let lockTimer: number | undefined
+
+const useAutoLock = ({ unlockApp, onAuthRequired }: UseAutoLockProps) => {
+  const appState = useRef(AppState.currentState)
+  const settingsLoadedFromStorage = useAppSelector((s) => s.settings.loadedFromStorage)
+  const isCameraOpen = useAppSelector((s) => s.app.isCameraOpen)
+  const isWalletUnlocked = useAppSelector((s) => s.wallet.isUnlocked)
+  const biometricsRequiredForAppAccess = useAppSelector((s) => s.settings.usesBiometrics)
+  const autoLockSeconds = useAppSelector((s) => s.settings.autoLockSeconds)
+  const dispatch = useAppDispatch()
+
+  const [isAppStateChangeCallbackRegistered, setIsAppStateChangeCallbackRegistered] = useState(false)
+  const [needsWalletUnlock, setNeedsWalletUnlock] = useState(false)
+
+  const lockApp = useCallback(() => {
+    if (biometricsRequiredForAppAccess) onAuthRequired()
+
+    dispatch(appBecameInactive())
+    // The following is needed when the switch between background/active happens so fast that the component didn't
+    // have enough time to re-render after clearning the mnemonic.
+    setNeedsWalletUnlock(true)
+  }, [biometricsRequiredForAppAccess, dispatch, onAuthRequired])
+
+  useEffect(() => {
+    if (!settingsLoadedFromStorage) return
+
+    const handleAppStateChange = (nextAppState: AppStateStatus) => {
+      if (autoLockSeconds !== -1) {
+        if (nextAppState === 'background' && isWalletUnlocked && !isCameraOpen) {
+          if (autoLockSeconds === 0) {
+            lockApp()
+          } else {
+            clearBackgroundTimer()
+            lockTimer = BackgroundTimer.setTimeout(() => {
+              if (lockTimer) {
+                lockApp()
+              }
+            }, autoLockSeconds * 1000)
+          }
+        } else if (nextAppState === 'active') {
+          clearBackgroundTimer()
+
+          if (!isWalletUnlocked && !isCameraOpen) {
+            setNeedsWalletUnlock(false)
+            unlockApp()
+          }
+        }
+      }
+
+      appState.current = nextAppState
+    }
+
+    if ((!isAppStateChangeCallbackRegistered || needsWalletUnlock) && appState.current === 'active') {
+      handleAppStateChange('active')
+    }
+
+    const subscription = AppState.addEventListener('change', handleAppStateChange)
+
+    setIsAppStateChangeCallbackRegistered(true)
+
+    return subscription.remove
+  }, [
+    autoLockSeconds,
+    biometricsRequiredForAppAccess,
+    dispatch,
+    isAppStateChangeCallbackRegistered,
+    isCameraOpen,
+    isWalletUnlocked,
+    lockApp,
+    needsWalletUnlock,
+    settingsLoadedFromStorage,
+    unlockApp
+  ])
+
+  const clearBackgroundTimer = () => {
+    if (!lockTimer) {
+      return
+    }
+    BackgroundTimer.clearTimeout(lockTimer)
+    lockTimer = undefined
+  }
+}
```

Comment:
Nit: could be a tiny bit shorter

```ts
const clearBackgroundTimer = () => {
  if (lockTimer) {
    BackgroundTimer.clearTimeout(lockTimer)
    lockTimer = undefined
  }
}
```

Reply:
true. I just copy+pasta from Metamask XD

Reply:
Fixed.